### ISSUE TITLE
s3cmd dies after computing md5 hashes

### ISSUE THREAD

**Issue by jwojcik** [s3tools/s3cmd#651]
### BEGIN Thu Oct 22 17:46:42 UTC 2015

[Thu Oct 22 17:46:42 UTC 2015] Starting sync of data to s3
INFO: No cache file found, creating it.
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Compiling list of local files...
INFO: Running stat() and reading/calculating MD5 values on 5784 files, this may take some time...
INFO: [1000/5784]
INFO: [2000/5784]
INFO: [3000/5784]
INFO: [4000/5784]
INFO: [5000/5784]

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    An unexpected error has occurred.
  Please try reproducing the error using
  the latest s3cmd code from the git master
  branch found at:
    https://github.com/s3tools/s3cmd
  and have a look at the known issues list:
    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions
  If the error persists, please report the
  following lines (removing any private
  info as necessary) to:
   s3tools-bugs@lists.sourceforge.net

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Invoked as: /usr/bin/s3cmd sync /var/lib/riak/leveldb/1004782375664995756265033322492444576013453623296 /var/lib/riak/leveldb/1050454301831586472458898473514828420377701515264 /var/lib/riak/leveldb/1096126227998177188652763624537212264741949407232 /var/lib/riak/leveldb/1141798154164767904846628775559596109106197299200 /var/lib/riak/leveldb/1187470080331358621040493926581979953470445191168 /var/lib/riak/leveldb/1233142006497949337234359077604363797834693083136 /var/lib/riak/leveldb/1278813932664540053428224228626747642198940975104 /var/lib/riak/leveldb/1324485858831130769622089379649131486563188867072 /var/lib/riak/leveldb/1370157784997721485815954530671515330927436759040 /var/lib/riak/leveldb/1415829711164312202009819681693899175291684651008 /var/lib/riak/leveldb/319703483166135013357056057156686910549735243776 /var/lib/riak/leveldb/365375409332725729550921208179070754913983135744 /var/lib/riak/leveldb/411047335499316445744786359201454599278231027712 /var/lib/riak/leveldb/456719261665907161938651510223838443642478919680 /var/lib/riak/leveldb/502391187832497878132516661246222288006726811648 /var/lib/riak/leveldb/548063113999088594326381812268606132370974703616 /var/lib/riak/leveldb/593735040165679310520246963290989976735222595584 /var/lib/riak/leveldb/639406966332270026714112114313373821099470487552 /var/lib/riak/leveldb/685078892498860742907977265335757665463718379520 /var/lib/riak/leveldb/730750818665451459101842416358141509827966271488 /var/lib/riak/leveldb/776422744832042175295707567380525354192214163456 /var/lib/riak/leveldb/822094670998632891489572718402909198556462055424 /var/lib/riak/leveldb/867766597165223607683437869425293042920709947392 /var/lib/riak/leveldb/913438523331814323877303020447676887284957839360 /var/lib/riak/leveldb/959110449498405040071168171470060731649205731328 s3://xxx-production-riak-backups/prod-riak-r04/2015.10.22/var/lib/riak/leveldb/ --cache-file=/tmp/sync2s3.md5 --exclude=*.log -H --verbose --delete-removed
Problem: KeyError: 'dev'
S3cmd:   1.6.0
python:   2.6.6 (r266:84292, Jan 22 2014, 09:42:36) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-4)]
environment LANG=en_US.UTF-8

Traceback (most recent call last):
  File "/usr/bin/s3cmd", line 2805, in <module>
    rc = main()
  File "/usr/bin/s3cmd", line 2713, in main
    rc = cmd_func(args)
  File "/usr/bin/s3cmd", line 1666, in cmd_sync
    return cmd_sync_local2remote(args)
  File "/usr/bin/s3cmd", line 1637, in cmd_sync_local2remote
    local_list, single_file_local, src_exclude_list = fetch_local_list(args[:-1], is_src = True, recursive = True)
  File "/usr/lib/python2.6/site-packages/S3/FileLists.py", line 342, in fetch_local_list
    _maintain_cache(cache, local_list)
  File "/usr/lib/python2.6/site-packages/S3/FileLists.py", line 298, in _maintain_cache
    cache.unmark_for_purge(local_list[i]['dev'], local_list[i]['inode'], local_list[i]['mtime'], local_list[i]['size'])
KeyError: 'dev'

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    An unexpected error has occurred.
  Please try reproducing the error using
  the latest s3cmd code from the git master
  branch found at:
    https://github.com/s3tools/s3cmd
  and have a look at the known issues list:
    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions
  If the error persists, please report the
  above lines (removing any private
  info as necessary) to:
   s3tools-bugs@lists.sourceforge.net
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[Thu Oct 22 19:42:30 UTC 2015] 0 files and directories now found at s3://xxx-production-riak-backups/prod-riak-r04/2015.10.22/var/lib/riak/leveldb

**Comment by fviard** [comment]
Thank you for your report.
I have a hypothesis, is it possible that some of the files are "deleted" during the time when md5 are calculated? Are you able to reproduce the issue consistently?

**Comment by jwojcik** [comment]
Yes, I am able to repeat the problem consistently.

Is it possible that a file is deleted while the md5 is being calculated?
Yes, it is possible. But in this case, it looks like the error happens
after all the md5s are calculated.

Odd thing is, it works fine on the other riak server. I'll try it on a
different server tomorrow and see if it works there.

**Comment by fviard** [comment]
What you find odd is in fact normal, because the code is failing in something like a "second pass" after the calculation of the md5s.

Ok, I was able to reproduce your issue, it really comes from a file vanishing between the "initial listing" step and the "md5 calculation" step.

I will need some time to find a better way to handle this case, but in the meantime, you can use the following patch if you want to be able to complete your backup job:
```diff
--- a/S3/FileLists.py
+++ b/S3/FileLists.py
@@ -298,6 +298,9 @@ def fetch_local_list(args, is_src = False, recursive = None):
         if cfg.cache_file and len(cfg.files_from) == 0:
             cache.mark_all_for_purge()
             for i in local_list.keys():
+                if 'dev' not in local_list[i] and 'inode' not in local_list and 'mtime' not in local_list:
+                    warning("File has vanished: %s", i)
+                    continue
                 cache.unmark_for_purge(local_list[i]['dev'], local_list[i]['inode'], local_list[i]['mtime'], local_list[i]['size'])
             cache.purge()
             cache.save(cfg.cache_file)
```

But, as a side note, i would like to warn you about doing such a backup of live data like database files. It is not a snapshot, during the time the backup take to complete from the first file to the last one, some files could be modified, resulting in uploaded data with files which are not coherent one with the other. 
But I'm not familiar with Riak or leveldb, so maybe their file structure is able to support such cases.

**Comment by jwojcik** [comment]
My approach is to backup the live database, then take the database down and
do a final sync. This way I can minimize the amount of time the db is down.
So, your warning is well taken, however I feel I have accounted for it.

John

**Comment by jwojcik** [comment]
The backups are working well. However, I am having a problem with detecting
a bad backup.

I throw an error in my monitoring system when the exit status of s3cmd is
!= 0.

The error I see in my logs is -
ERROR: Upload of <redacted> is not possible (Reason: Not a regular file)

When I look after the backup, I don't see the file on the filesystem. My
suspicion is that s3cmd builds the list of files to back up upon start, but
then when it gets around to back up certain files, the file may have been
deleted in the meantime.

It would be helpful to get an exit status specifically for the case where
files were deleted after the file list was built.

**Comment by jwojcik** [comment]
Bump

**Comment by qris** [comment]
Hi, I'm having the same issue backing up my Owncloud data store to S3. It seems to be repeatable, so it's possible that Owncloud is deleting files during some maintenance process, or clients are. (It's hard to tell because we don't get the name of the file that disappeared).

I'm not too fussed if some files are deleted (by clients) during the backup or some metadata is missing, I just really do not want to lose the latest version of my documents in Owncloud if my house burns down.

A fix would be very much appreciated :)

**Comment by fviard** [comment]
Closing as the issue should already be fixed in release 2.1.0.
